File: src/VoidRay/Services/LinkParser.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using VoidRay.Models;

namespace VoidRay.Services;

/// <summary>Parses share links (vless, vmess, trojan, ss) and subscription bodies.</summary>
public static class LinkParser
{
    public static List<ServerProfile> ParseSubscriptionBody(string body, out int unsupported)
    {
        unsupported = 0;
        var result = new List<ServerProfile>();
        if (string.IsNullOrWhiteSpace(body))
            return result;

        var text = body.Trim();

        // Some panels (Marzban/Remnawave custom JSON) return full Xray configs.
        if (text.StartsWith('[') || text.StartsWith('{'))
        {
            result.AddRange(ParseXrayJson(text));
            if (result.Count > 0)
                return result;
        }

        if (!text.Contains("://"))
        {
            var decoded = Base64.TryDecode(text);
            if (decoded is not null && decoded.Contains("://"))
                text = decoded.Trim();
        }

        foreach (var line in text.Split(new[] { '\r', '\n' },
                     StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            if (!line.Contains("://"))
                continue;
            var profile = TryParse(line);
            if (profile is null)
            {
                unsupported++;
            }
            else
            {
                profile.RawLink = line;
                result.Add(profile);
            }
        }
        return result;
    }

    public static ServerProfile? TryParse(string link)
    {
        try
        {
            var idx = link.IndexOf("://", StringComparison.Ordinal);
            if (idx <= 0)
                return null;
            return link[..idx].ToLowerInvariant() switch
            {
                "vless" => ParseVlessOrTrojan(link, "vless"),
                "trojan" => ParseVlessOrTrojan(link, "trojan"),
                "vmess" => ParseVmess(link),
                "ss" => ParseShadowsocks(link),
                _ => null,
            };
        }
        catch
        {
            return null;
        }
    }

    // ---------------------------------------------------------------- vless / trojan

    private static ServerProfile? ParseVlessOrTrojan(string link, string protocol)
    {
        var u = UrlParts.Parse(link);
        if (u is null || string.IsNullOrEmpty(u.UserInfo))
            return null;

        var p = new ServerProfile
        {
            Protocol = protocol,
            Id = u.UserInfo,
            Address = u.Host,
            Port = u.Port,
            Remark = string.IsNullOrWhiteSpace(u.Fragment) ? $"{u.Host}:{u.Port}" : u.Fragment,
        };
        FillTransport(p, u.Query);
        if (protocol == "trojan" && !u.Query.ContainsKey("security"))
            p.Security = "tls";
        return p;
    }

    private static void FillTransport(ServerProfile p, IReadOnlyDictionary<string, string> q)
    {
        string? Get(string k) => q.TryGetValue(k, out var v) && !string.IsNullOrEmpty(v) ? v : null;

        p.Network = Get("type") ?? "tcp";
        p.Security = Get("security") ?? "none";
        p.Sni = Get("sni") ?? Get("peer");
        p.Fingerprint = Get("fp");
        p.Alpn = Get("alpn");
        p.PublicKey = Get("pbk");
        p.ShortId = Get("sid");
        p.SpiderX = Get("spx");
        p.Path = Get("path");
        p.Host = Get("host");
        p.ServiceName = Get("serviceName");
        p.Mode = Get("mode");
        p.HeaderType = Get("headerType");
        p.Flow = Get("flow");
        p.Encryption = Get("encryption");
        p.Extra = Get("extra");
        p.AllowInsecure = Get("allowInsecure") is "1" or "true" || Get("insecure") is "1" or "true";
    }

    // ---------------------------------------------------------------- vmess

    private static ServerProfile? ParseVmess(string link)
    {
        var payload = link[8..];
        var hash = payload.IndexOf('#');
        if (hash >= 0)
            payload = payload[..hash];
        var json = Base64.TryDecode(payload);
        if (json is null)
            return null;

        using var doc = JsonDocument.Parse(json);
        var r = doc.RootElement;
        string? S(string name) =>
            r.TryGetProperty(name, out var v)
                ? v.ValueKind switch
                {
                    JsonValueKind.String => string.IsNullOrEmpty(v.GetString()) ? null : v.GetString(),
                    JsonValueKind.Number => v.GetRawText(),
                    _ => null,
                }
                : null;

        var address = S("add");
        if (address is null || !int.TryParse(S("port"), out var port))
            return null;

        return new ServerProfile
        {
            Protocol = "vmess",
            Remark = S("ps") ?? $"{address}:{port}",
            Address = address,
            Port = port,
            Id = S("id"),
            AlterId = int.TryParse(S("aid"), out var aid) ? aid : 0,
            VmessSecurity = S("scy") ?? "auto",
            Network = S("net") ?? "tcp",
            HeaderType = S("type"),
            Host = S("host"),
            Path = S("path"),
            ServiceName = S("net") == "grpc" ? S("path") : null,
            Security = S("tls") ?? "none",
            Sni = S("sni"),
            Alpn = S("alpn"),
            Fingerprint = S("fp"),
        };
    }

    // ---------------------------------------------------------------- shadowsocks

    private static ServerProfile? ParseShadowsocks(string link)
    {
        var rest = link[5..];
        var fragment = "";
        var hash = rest.IndexOf('#');
        if (hash >= 0)
        {
            fragment = UrlParts.Unescape(rest[(hash + 1)..]);
            rest = rest[..hash];
        }
        var q = rest.IndexOf('?');
        if (q >= 0)
            rest = rest[..q];
        rest = rest.TrimEnd('/');

        string userInfo, hostPort;
        var at = rest.LastIndexOf('@');
        if (at >= 0)
        {
            var raw = rest[..at];
            userInfo = Base64.TryDecode(raw) is { } dec && dec.Contains(':') ? dec : UrlParts.Unescape(raw);
            hostPort = rest[(at + 1)..];
        }
        else
        {
            var decoded = Base64.TryDecode(rest);
            if (decoded is null)
                return null;
            at = decoded.LastIndexOf('@');
            if (at < 0)
                return null;
            userInfo = decoded[..at];
            hostPort = decoded[(at + 1)..];
        }

        var colon = userInfo.IndexOf(':');
        if (colon < 0 || !UrlParts.TrySplitHostPort(hostPort, out var host, out var port))
            return null;

        return new ServerProfile
        {
            Protocol = "shadowsocks",
            Method = userInfo[..colon],
            Id = userInfo[(colon + 1)..],
            Address = host,
            Port = port,
            Remark = string.IsNullOrWhiteSpace(fragment) ? $"{host}:{port}" : fragment,
        };
    }

    // ---------------------------------------------------------------- xray json

    private static IEnumerable<ServerProfile> ParseXrayJson(string text)
    {
        JsonNode? root;
        try { root = JsonNode.Parse(text); }
        catch { yield break; }

        var configs = root switch
        {
            JsonArray arr => arr.OfType<JsonObject>(),
            JsonObject obj => new[] { obj },
            _ => Enumerable.Empty<JsonObject>(),
        };

        foreach (var cfg in configs)
        {
            if (cfg["outbounds"] is not JsonArray outbounds)
                continue;
            var proxy = outbounds.OfType<JsonObject>()
                .FirstOrDefault(o => (string?)o["protocol"] is "vless" or "vmess" or "trojan" or "shadowsocks");
            if (proxy is null)
                continue;

            var settings = proxy["settings"] as JsonObject;
            var server = (settings?["vnext"] as JsonArray)?[0] as JsonObject
                         ?? (settings?["servers"] as JsonArray)?[0] as JsonObject
                         ?? settings;
            var address = (string?)server?["address"] ?? "";
            var port = server?["port"] is JsonValue pv && pv.TryGetValue<int>(out var pi) ? pi : 0;
            var stream = proxy["streamSettings"] as JsonObject;

            yield return new ServerProfile
            {
                Protocol = (string)proxy["protocol"]!,
                Remark = (string?)cfg["remarks"] ?? $"{address}:{port}",
                Address = address,
                Port = port,
                Network = (string?)stream?["network"] ?? "tcp",
                Security = (string?)stream?["security"] ?? "none",
                RawXrayJson = cfg.ToJsonString(),
            };
        }
    }
}

internal sealed class UrlParts
{
    public string UserInfo { get; private init; } = "";
    public string Host { get; private init; } = "";
    public int Port { get; private init; }
    public string Fragment { get; private init; } = "";
    public Dictionary<string, string> Query { get; } = new(StringComparer.OrdinalIgnoreCase);

    public static UrlParts? Parse(string link)
    {
        var idx = link.IndexOf("://", StringComparison.Ordinal);
        var rest = link[(idx + 3)..];

        var fragment = "";
        var hash = rest.IndexOf('#');
        if (hash >= 0)
        {
            fragment = Unescape(rest[(hash + 1)..]);
            rest = rest[..hash];
        }

        var queryString = "";
        var q = rest.IndexOf('?');
        if (q >= 0)
        {
            queryString = rest[(q + 1)..];
            rest = rest[..q];
        }
        rest = rest.TrimEnd('/');

        var user = "";
        var at = rest.LastIndexOf('@');
        if (at >= 0)
        {
            user = Unescape(rest[..at]);
            rest = rest[(at + 1)..];
        }

        if (!TrySplitHostPort(rest, out var host, out var port))
            return null;

        var parts = new UrlParts { UserInfo = user, Host = host, Port = port, Fragment = fragment };
        foreach (var pair in queryString.Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var eq = pair.IndexOf('=');
            if (eq < 0)
                parts.Query[Unescape(pair)] = "";
            else
                parts.Query[Unescape(pair[..eq])] = Unescape(pair[(eq + 1)..]);
        }
        return parts;
    }

    public static bool TrySplitHostPort(string value, out string host, out int port)
    {
        host = "";
        port = 0;
        if (value.StartsWith('['))
        {
            var close = value.IndexOf(']');
            if (close < 0)
                return false;
            host = value[1..close];
            return value.Length > close + 2 && int.TryParse(value[(close + 2)..], out port);
        }
        var colon = value.LastIndexOf(':');
        if (colon <= 0)
            return false;
        host = value[..colon];
        return int.TryParse(value[(colon + 1)..], out port);
    }

    public static string Unescape(string s)
    {
        try { return Uri.UnescapeDataString(s); }
        catch { return s; }
    }
}

public static class Base64
{
    public static string? TryDecode(string input)
    {
        var sb = new StringBuilder(input.Length + 3);
        foreach (var c in input)
        {
            if (char.IsWhiteSpace(c))
                continue;
            sb.Append(c switch { '-' => '+', '_' => '/', _ => c });
        }
        var t = sb.ToString().TrimEnd('=');
        switch (t.Length % 4)
        {
            case 1: return null;
            case 2: t += "=="; break;
            case 3: t += "="; break;
        }
        try
        {
            return Encoding.UTF8.GetString(Convert.FromBase64String(t));
        }
        catch
        {
            return null;
        }
    }
}

File: android/app/src/main/java/win/enderr/voidray/core/LinkParser.kt
package win.enderr.voidray.core

import org.json.JSONArray
import org.json.JSONObject
import java.net.URLDecoder

/** Parses share links (vless, vmess, trojan, ss) and subscription bodies. */
object LinkParser {

    class Result(val servers: List<ServerProfile>, val unsupported: Int)

    fun parseSubscriptionBody(body: String): Result {
        val servers = mutableListOf<ServerProfile>()
        var unsupported = 0
        var text = body.trim()
        if (text.isEmpty()) return Result(servers, 0)

        // Some panels return full Xray configs.
        if (text.startsWith("[") || text.startsWith("{")) {
            servers += parseXrayJson(text)
            if (servers.isNotEmpty()) return Result(servers, 0)
        }

        if (!text.contains("://")) {
            val decoded = Base64Util.tryDecode(text)
            if (decoded != null && decoded.contains("://")) text = decoded.trim()
        }

        for (raw in text.split('\r', '\n')) {
            val line = raw.trim()
            if (!line.contains("://")) continue
            val profile = tryParse(line)
            if (profile == null) unsupported++ else servers += profile.also { it.rawLink = line }
        }
        return Result(servers, unsupported)
    }

    fun tryParse(link: String): ServerProfile? = try {
        when (link.substringBefore("://").lowercase()) {
            "vless" -> parseVlessOrTrojan(link, "vless")
            "trojan" -> parseVlessOrTrojan(link, "trojan")
            "vmess" -> parseVmess(link)
            "ss" -> parseShadowsocks(link)
            else -> null
        }
    } catch (e: Exception) {
        null
    }

    // ------------------------------------------------------------ vless / trojan

    private fun parseVlessOrTrojan(link: String, protocol: String): ServerProfile? {
        val u = UrlParts.parse(link) ?: return null
        if (u.userInfo.isEmpty()) return null
        val p = ServerProfile(
            protocol = protocol,
            id = u.userInfo,
            address = u.host,
            port = u.port,
            remark = u.fragment.ifBlank { "${u.host}:${u.port}" },
        )
        fillTransport(p, u.query)
        if (protocol == "trojan" && !u.query.containsKey("security")) p.security = "tls"
        return p
    }

    private fun fillTransport(p: ServerProfile, q: Map<String, String>) {
        fun get(k: String) = q[k]?.takeIf { it.isNotEmpty() }
        p.network = get("type") ?: "tcp"
        p.security = get("security") ?: "none"
        p.sni = get("sni") ?: get("peer")
        p.fingerprint = get("fp")
        p.alpn = get("alpn")
        p.publicKey = get("pbk")
        p.shortId = get("sid")
        p.spiderX = get("spx")
        p.path = get("path")
        p.host = get("host")
        p.serviceName = get("serviceName")
        p.mode = get("mode")
        p.headerType = get("headerType")
        p.flow = get("flow")
        p.encryption = get("encryption")
        p.extra = get("extra")
        p.allowInsecure = get("allowInsecure") in setOf("1", "true") || get("insecure") in setOf("1", "true")
    }

    // ------------------------------------------------------------ vmess

    private fun parseVmess(link: String): ServerProfile? {
        val payload = link.substring(8).substringBefore('#')
        val json = Base64Util.tryDecode(payload) ?: return null
        val o = JSONObject(json)
        fun s(name: String): String? = if (!o.has(name) || o.isNull(name)) null else o.get(name).toString().takeIf { it.isNotEmpty() }
        val address = s("add") ?: return null
        val port = s("port")?.toIntOrNull() ?: return null
        return ServerProfile(
            protocol = "vmess",
            remark = s("ps") ?: "$address:$port",
            address = address,
            port = port,
            id = s("id"),
            alterId = s("aid")?.toIntOrNull() ?: 0,
            vmessSecurity = s("scy") ?: "auto",
            network = s("net") ?: "tcp",
            headerType = s("type"),
            host = s("host"),
            path = s("path"),
            serviceName = if (s("net") == "grpc") s("path") else null,
            security = s("tls") ?: "none",
            sni = s("sni"),
            alpn = s("alpn"),
            fingerprint = s("fp"),
        )
    }

    // ------------------------------------------------------------ shadowsocks

    private fun parseShadowsocks(link: String): ServerProfile? {
        var rest = link.substring(5)
        var fragment = ""
        rest.indexOf('#').takeIf { it >= 0 }?.let {
            fragment = UrlParts.unescape(rest.substring(it + 1))
            rest = rest.substring(0, it)
        }
        rest = rest.substringBefore('?').trimEnd('/')

        val userInfo: String
        val hostPort: String
        val at = rest.lastIndexOf('@')
        if (at >= 0) {
            val raw = rest.substring(0, at)
            val dec = Base64Util.tryDecode(raw)
            userInfo = if (dec != null && dec.contains(':')) dec else UrlParts.unescape(raw)
            hostPort = rest.substring(at + 1)
        } else {
            val decoded = Base64Util.tryDecode(rest) ?: return null
            val at2 = decoded.lastIndexOf('@')
            if (at2 < 0) return null
            userInfo = decoded.substring(0, at2)
            hostPort = decoded.substring(at2 + 1)
        }
        val colon = userInfo.indexOf(':')
        val hp = UrlParts.splitHostPort(hostPort)
        if (colon < 0 || hp == null) return null
        return ServerProfile(
            protocol = "shadowsocks",
            method = userInfo.substring(0, colon),
            id = userInfo.substring(colon + 1),
            address = hp.first,
            port = hp.second,
            remark = fragment.ifBlank { "${hp.first}:${hp.second}" },
        )
    }

    // ------------------------------------------------------------ xray json

    private fun parseXrayJson(text: String): List<ServerProfile> {
        val configs = try {
            when (val root = if (text.startsWith("[")) JSONArray(text) else JSONObject(text)) {
                is JSONArray -> (0 until root.length()).mapNotNull { root.optJSONObject(it) }
                is JSONObject -> listOf(root)
                else -> emptyList()
            }
        } catch (e: Exception) {
            return emptyList()
        }

        val result = mutableListOf<ServerProfile>()
        for (cfg in configs) {
            val outbounds = cfg.optJSONArray("outbounds") ?: continue
            val proxy = (0 until outbounds.length()).mapNotNull { outbounds.optJSONObject(it) }
                .firstOrNull { it.optString("protocol") in setOf("vless", "vmess", "trojan", "shadowsocks") } ?: continue
            val settings = proxy.optJSONObject("settings")
            val server = settings?.optJSONArray("vnext")?.optJSONObject(0)
                ?: settings?.optJSONArray("servers")?.optJSONObject(0)
                ?: settings
            val address = server?.optString("address").orEmpty()
            val port = server?.optInt("port") ?: 0
            val stream = proxy.optJSONObject("streamSettings")
            result += ServerProfile(
                protocol = proxy.getString("protocol"),
                remark = cfg.str("remarks") ?: "$address:$port",
                address = address,
                port = port,
                network = stream?.str("network") ?: "tcp",
                security = stream?.str("security") ?: "none",
                rawXrayJson = cfg.toString(),
            )
        }
        return result
    }
}

internal class UrlParts(
    val userInfo: String,
    val host: String,
    val port: Int,
    val fragment: String,
    val query: Map<String, String>,
) {
    companion object {
        fun parse(link: String): UrlParts? {
            var rest = link.substringAfter("://")
            var fragment = ""
            rest.indexOf('#').takeIf { it >= 0 }?.let {
                fragment = unescape(rest.substring(it + 1))
                rest = rest.substring(0, it)
            }
            var queryString = ""
            rest.indexOf('?').takeIf { it >= 0 }?.let {
                queryString = rest.substring(it + 1)
                rest = rest.substring(0, it)
            }
            rest = rest.trimEnd('/')

            var user = ""
            val at = rest.lastIndexOf('@')
            if (at >= 0) {
                user = unescape(rest.substring(0, at))
                rest = rest.substring(at + 1)
            }
            val (host, port) = splitHostPort(rest) ?: return null

            val query = sortedMapOf<String, String>(String.CASE_INSENSITIVE_ORDER)
            for (pair in queryString.split('&')) {
                if (pair.isEmpty()) continue
                val eq = pair.indexOf('=')
                if (eq < 0) query[unescape(pair)] = ""
                else query[unescape(pair.substring(0, eq))] = unescape(pair.substring(eq + 1))
            }
            return UrlParts(user, host, port, fragment, query)
        }

        fun splitHostPort(value: String): Pair<String, Int>? {
            if (value.startsWith("[")) {
                val close = value.indexOf(']')
                if (close < 0 || value.length <= close + 2) return null
                val port = value.substring(close + 2).toIntOrNull() ?: return null
                return value.substring(1, close) to port
            }
            val colon = value.lastIndexOf(':')
            if (colon <= 0) return null
            val port = value.substring(colon + 1).toIntOrNull() ?: return null
            return value.substring(0, colon) to port
        }

        /** Percent-decoding that keeps '+' as is (share links are not form-encoded). */
        fun unescape(s: String): String = try {
            URLDecoder.decode(s.replace("+", "%2B"), "UTF-8")
        } catch (e: Exception) {
            s
        }
    }
}
